updated to nodejs10.x after AWS deprecated nodejs8.10

agstevens opened the pull request:
and updated site-builder to mime2.0.5 so it deploys. the lookup() method in mime had been deprecated in favor of getType() in mime versions > 2

agstevens commented:
From the AWS discontinuation notice.

For Node.js 8.x, there will be 2 stages to the runtime deprecation process:

Disable Function Create – Beginning January 6, 2020, customers will no longer be able to create functions using Node.js 8.10

Disable Function Update – Beginning February 3, 2020, customers will no longer be able to update functions using Node.js 8.10

agstevens commented:
Canceling this pull request in favor of a new one. These changes made it deploy, but there was an issue with ImageMagick which I have since fixed and the fix will be included in the next pull request.

agstevens commented:
Added in the ImageMagick fix using a Lambda layer. 
Updated the documentation. 
Tested only with Node.js 10.x.

tgrecojr commented:
Thank you so much for this PR.  I would have been lost with the ImageMagick removal from node in this version.  I hope it gets merged soon.

agstevens commented:
I got it working, but when I uploaded a larger number of images, I saw some higher than expected charges.

I think the expense is 90% the s3 read/write requests from rebuilding the site on every single image resize upload and 10% cloudfront invalidations. I'm reworking it to only run the site builder on a schedule (every few hours), instead of in response to each resized image being uploaded: 

https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/RunLambdaSchedule.html

Will test with a smaller amount of images and verify that it works.

agstevens commented:
I removed the lambda trigger from the resized bucket that was causing the sitebuilder to run something like O[2N^2] S3 PUT, LIST and GET requests, which can start to add up.

I replaced it with a CloudWatch Events regularly timed event that defaults to 365 days. This results in a much more modest number, I estimate  O[6N] of S3 PUT, LIST and GET requests. 

I also added:
- Serverside AES 256 encryption to the photo buckets, and forced encryption policies to make sure you upload photos with the encryption enabled. CloudFront can automatically decrypt them, but they are encrypted at rest now.
- Public Access Block Configuration for the photo buckets - since you're serving through cloudfront, you just don't need those to be publicly accessible.
- CloudFront longest cache times possible (possibly useful for a relatively static site) and compression if the browser supports it.
- Set CloudFront price class to the cheapest option.
- Fixed a typo in sitebuilder for the declaration of the comment2 variable.

agstevens commented:
I removed the Public Access Block Configuration - wasn't passing my testing. Instead of setting it before, I think it is best to set it after the site is built and deployed and all the photos have been uploaded. I also like to throttle the lambda functions (sitebuilder and resize) when I know they will not be used, and they can be enabled if you decide to add more photos later.

jpsim commented:
This is awesome @agstevens. I especially appreciate the time you put into writing careful documentation and sharing more context here.

I wish all open source contributors were as caring as you 🥇
